File: HelloWorld/HelloWorldObjectiveC/HelloWorldObjectiveC/ViewController.m
//
//  ViewController.m
//  HelloWorldObjectiveC
//
//  Created by Eric Lange on 8/19/18.
//  Copyright © 2018 LiquidPlayer. All rights reserved.
//

#import "ViewController.h"
#import <LiquidCore/LiquidCore.h>

@interface ViewController ()  <LCMicroServiceDelegate, LCMicroServiceEventListener>

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.text = [[UILabel alloc] init];
    self.text.textAlignment = NSTextAlignmentCenter;
    self.text.text = @"Hello World!";
    self.text.font = [UIFont fontWithName:@"Menlo" size:17.];
    [self.view addSubview:self.text];
    
    self.button = [UIButton buttonWithType:UIButtonTypeSystem];
    [self.button setTitle:@"Sprechen Sie Deutsch!" forState:UIControlStateNormal];
    [self.button.titleLabel setFont:[UIFont fontWithName:@"Menlo" size:17.]];
    [self.button addTarget:self action:@selector(onTouch:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:self.button];

    self.text.translatesAutoresizingMaskIntoConstraints = NO;
    self.button.translatesAutoresizingMaskIntoConstraints = NO;
    
    UILayoutGuide *topGuide = [[UILayoutGuide alloc] init];
    [self.view addLayoutGuide:topGuide];
    
    UILayoutGuide *bottomGuide = [[UILayoutGuide alloc] init];
    [self.view addLayoutGuide:bottomGuide];
    
    id views = @{
                 @"text": self.text,
                 @"button": self.button,
                 @"topGuide": topGuide,
                 @"bottomGuide": bottomGuide,
                 };
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-[text]-|"
                                                                      options:0 metrics:nil views:views]];
    [self.view addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|-[button]-|"
                                                                      options:0 metrics:nil views:views]];
    [self.view addConstraints:[NSLayoutConstraint
                               constraintsWithVisualFormat:@"V:|[topGuide]-[text]-[button]-[bottomGuide(==topGuide)]|"
                               options:0 metrics:nil views:views]];
}

- (void) onTouch:(UIButton*)sender
{
    NSURL *url = [LCMicroService devServer];
    LCMicroService *service = [[LCMicroService alloc] initWithURL:url delegate:self];
    [service start];
}

- (void) onStart:(LCMicroService *)service
{
    [service addEventListener:@"ready" listener:self];
    [service addEventListener:@"pong" listener:self];
}

- (void) onEvent:(LCMicroService *)service event:(NSString *)event payload:(id)payload
{
    if ([event isEqualToString:@"ready"]) {
        [service emit:@"ping"];
    } else if ([event isEqualToString:@"pong"]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.text setText:payload[@"message"]];
        });
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: HelloWorld/HelloWorldSwift/HelloWorldSwift/ViewController.swift
//
//  ViewController.swift
//  HelloWorldSwift
//
//  Created by Eric Lange on 8/22/18.
//  Copyright © 2018 LiquidPlayer. All rights reserved.
//

import UIKit
import LiquidCore

class ViewController: UIViewController, LCMicroServiceDelegate, LCMicroServiceEventListener {

    var text: UILabel = UILabel()
    var button: UIButton = UIButton(type: .system)
    
    override func viewDidLoad() {
        super.viewDidLoad()

        text.textAlignment = .center
        text.text = "Hello World!"
        text.font = UIFont(name: "Menlo", size: 17)
        self.view.addSubview(text)
        
        button.setTitle("Sprechen Sie Deutsch!", for: .normal)
        button.titleLabel?.font = UIFont(name: "Menlo", size: 17)
        button.addTarget(self, action: #selector(onTouch), for: .touchUpInside)
        self.view.addSubview(button)
        
        self.text.translatesAutoresizingMaskIntoConstraints = false
        self.button.translatesAutoresizingMaskIntoConstraints = false
        
        let top = UILayoutGuide()
        let bottom = UILayoutGuide()
        self.view.addLayoutGuide(top)
        self.view.addLayoutGuide(bottom)
        
        let views = [ "text": text, "button": button, "top":top, "bottom":bottom ]
        let c1 = NSLayoutConstraint.constraints(withVisualFormat: "H:|-[text]-|", metrics: nil, views: views)
        let c2 = NSLayoutConstraint.constraints(withVisualFormat: "H:|-[button]-|", metrics: nil, views: views)
        let c3 = NSLayoutConstraint.constraints(withVisualFormat: "V:|[top]-[text]-[button]-[bottom(==top)]|",
                                                metrics: nil, views: views)
        self.view.addConstraints(c1 + c2 + c3)
    }
    
    @objc func onTouch(sender:UIButton!) {
        let url = LCMicroService.devServer()
        let service = LCMicroService(url: url, delegate: self)
        service?.start()
    }
    
    func onStart(_ service: LCMicroService) {
        service.addEventListener("ready", listener: self)
        service.addEventListener("pong", listener: self)
    }
    
    func onEvent(_ service: LCMicroService, event: String, payload: Any?) {
        if event == "ready" {
            service.emit("ping")
        } else if event == "pong" {
            DispatchQueue.main.async {
                var p = (payload as! Dictionary<String,AnyObject>)
                self.text.text = p["message"] as? String
            }
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

